### DISCUSSION TITLE
CMake auto add classes files

### DISCUSSION THREAD

**Discussion by DelinWorks** [axmolengine/axmol#626] [Q&A] [answered]
Hi! is there a way to automatically add `.c`, `.cpp`, `.h` and `.hpp` files to my project in cmake?
This is my project's structure:

![image](https://user-images.githubusercontent.com/45469625/165995055-fe1d6912-e2a2-4a66-b25d-f2439df891f0.png)

and I'm thinking of dividing `GameUtils.h` 

![image](https://user-images.githubusercontent.com/45469625/165995585-461a31d5-f473-4737-8562-74b5d8364796.png)

But it's too much hassle to go to the cmake file and add every file
It would be awesome if you could specify something like `Classes/*`

If there is nothing like this It's okay because I've got a workaround for this in C#

```
using System;

public class main
{
    public static void printRecursive(string filter)
    {
        foreach (string file in Directory.EnumerateFiles("D:/Development/Projects/DarknessAmongUs/Classes", filter, SearchOption.AllDirectories))
        {
            Console.WriteLine("    " + file.Replace("D:/Development/Projects/DarknessAmongUs/", "").Replace("\\", "/"));
        }
    }

    public static void Main()
    {
        Console.WriteLine("list(APPEND GAME_SOURCE");

        printRecursive("*.c");
        Console.WriteLine();

        printRecursive("*.cpp");
        Console.WriteLine();

        Console.WriteLine(")\n");

        Console.WriteLine("list(APPEND GAME_HEADER");

        printRecursive("*.h");
        Console.WriteLine();

        printRecursive("*.hpp");
        Console.WriteLine();

        Console.WriteLine(")\n");

        Console.WriteLine("\nCopy this into your cmake file.\n");

    }
}
```

**Comment by halx99** [comment, marked as answer]
Yes, cmake support it

**Reply by halx99** [reply]
please refer to https://github.com/adxeproject/adxe/blob/dev/extensions/fairygui/CMakeLists.txt

**Reply by DelinWorks** [reply]
Thanks! it worked really appreciate it ❤

**Comment by rh101** [comment]
Be careful when using GLOB_RECURSE, because if I recall correctly, it won't detect new files added to your project, and would require the actual CMakeLists.txt to be modified in some way (or simply last modified date changed) in order for it to do a re-build.

Read up here for more info: [https://stackoverflow.com/questions/1027247/is-it-better-to-specify-source-files-with-glob-or-each-file-individually-in-cmak/1060061#1060061](https://stackoverflow.com/questions/1027247/is-it-better-to-specify-source-files-with-glob-or-each-file-individually-in-cmak/1060061#1060061)

Many other posts online about it too.

Also, you can add a separate CMakeLists.txt in each sub-folder, and reference them from the parent folder etc.  That way it keeps your root CMakeLists.txt clean.  It may seem like a hassle to add each new source file into CMakeLists.txt manually, but it only ever needs to be done once per file, and you never have to deal with it for that file again.  

Just as an example, my largest project has over 900 source files and 36 folders just in the "Classes" folder alone, and they're all listed explicitly in a number of CMakeLists.txt files, without any usage of GLOB etc.  It takes a few seconds to add the cpp/h file reference into whichever CMakeLists.txt it belongs in, and that's it, it's done.

**Reply by DelinWorks** [reply]
Yeah i noticed it, i just open it in notepad add a space and then remove and then save and it automatically rebuilds. 
that's actually a good idea, separating folders with cmakelists it feels like it would make a project clean, Thanks! will look into it.

**Reply by rh101** [reply]
Instead of modifying the file and re-saving, you can just run `touch` on the file if you're using a Linux based environment, or on Windows, something like what is suggested here: [https://superuser.com/questions/10426/windows-equivalent-of-the-linux-command-touch](https://superuser.com/questions/10426/windows-equivalent-of-the-linux-command-touch)

That is assuming you want to stick with GLOB.

**Reply by DelinWorks** [reply]
never thought of that 🤔 Thanks!
